File: data_baseline/openrouter__qwen__qwen3.7-plus/java/rep2/Main.java
import java.io.*;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.time.*;
import java.time.format.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import com.sun.net.httpserver.*;

public class Main {
    static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'").withZone(ZoneOffset.UTC);
    
    static String now() {
        return FORMATTER.format(Instant.now());
    }

    static Map<String, Object> mapOf(Object... args) {
        Map<String, Object> map = new LinkedHashMap<>();
        for (int i = 0; i < args.length; i += 2) {
            map.put((String) args[i], args[i + 1]);
        }
        return map;
    }

    static class User {
        int id;
        String username;
        String password;
        User(int id, String username, String password) {
            this.id = id;
            this.username = username;
            this.password = password;
        }
    }

    static class Todo {
        int id;
        int userId;
        String title;
        String description;
        boolean completed;
        String createdAt;
        String updatedAt;
        
        Map<String, Object> toJson() {
            Map<String, Object> map = new LinkedHashMap<>();
            map.put("id", id);
            map.put("title", title);
            map.put("description", description);
            map.put("completed", completed);
            map.put("created_at", createdAt);
            map.put("updated_at", updatedAt);
            return map;
        }
    }

    static final AtomicInteger nextUserId = new AtomicInteger(1);
    static final AtomicInteger nextTodoId = new AtomicInteger(1);
    static final Map<String, User> usersByUsername = new ConcurrentHashMap<>();
    static final Map<Integer, User> usersById = new ConcurrentHashMap<>();
    static final Map<String, Integer> sessions = new ConcurrentHashMap<>();
    static final Map<Integer, Todo> todos = new ConcurrentHashMap<>();

    static String generateToken() {
        return UUID.randomUUID().toString().replace("-", "");
    }

    static Integer getUserId(HttpExchange exchange) {
        String cookieHeader = exchange.getRequestHeaders().getFirst("Cookie");
        if (cookieHeader != null) {
            String[] cookies = cookieHeader.split(";");
            for (String cookie : cookies) {
                cookie = cookie.trim();
                if (cookie.startsWith("session_id=")) {
                    String token = cookie.substring("session_id=".length()).trim();
                    return sessions.get(token);
                }
            }
        }
        return null;
    }

    static String readBody(HttpExchange exchange) throws IOException {
        InputStream is = exchange.getRequestBody();
        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        int nRead;
        byte[] data = new byte[1024];
        while ((nRead = is.read(data, 0, data.length)) != -1) {
            buffer.write(data, 0, nRead);
        }
        return buffer.toString(StandardCharsets.UTF_8.name());
    }

    static void sendJson(HttpExchange exchange, int status, Object data) throws IOException {
        String json = JsonParser.stringify(data);
        byte[] bytes = json.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(status, bytes.length);
        try (OutputStream os = exchange.getResponseBody()) {
            os.write(bytes);
        }
    }

    static void sendNoContent(HttpExchange exchange, int status) throws IOException {
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(status, -1);
    }

    static void handleRegister(HttpExchange exchange) throws IOException {
        if (!exchange.getRequestMethod().equals("POST")) {
            sendJson(exchange, 405, mapOf("error", "Method not allowed"));
            return;
        }
        String body = readBody(exchange);
        Map<String, Object> req;
        try {
            req = JsonParser.parseObject(body);
        } catch (Exception e) {
            sendJson(exchange, 400, mapOf("error", "Invalid request"));
            return;
        }
        
        Object usernameObj = req.get("username");
        Object passwordObj = req.get("password");
        
        if (!(usernameObj instanceof String) || !(passwordObj instanceof String)) {
            sendJson(exchange, 400, mapOf("error", "Invalid request"));
            return;
        }
        String username = (String) usernameObj;
        String password = (String) passwordObj;
        
        if (!username.matches("^[a-zA-Z0-9_]{3,50}$")) {
            sendJson(exchange, 400, mapOf("error", "Invalid username"));
            return;
        }
        if (password.length() < 8) {
            sendJson(exchange, 400, mapOf("error", "Password too short"));
            return;
        }
        
        synchronized (usersByUsername) {
            if (usersByUsername.containsKey(username)) {
                sendJson(exchange, 409, mapOf("error", "Username already exists"));
                return;
            }
            int id = nextUserId.getAndIncrement();
            User user = new User(id, username, password);
            usersByUsername.put(username, user);
            usersById.put(id, user);
            sendJson(exchange, 201, mapOf("id", id, "username", username));
        }
    }

    static void handleLogin(HttpExchange exchange) throws IOException {
        if (!exchange.getRequestMethod().equals("POST")) {
            sendJson(exchange, 405, mapOf("error", "Method not allowed"));
            return;
        }
        String body = readBody(exchange);
        Map<String, Object> req;
        try {
            req = JsonParser.parseObject(body);
        } catch (Exception e) {
            sendJson(exchange, 400, mapOf("error", "Invalid request"));
            return;
        }
        
        Object usernameObj = req.get("username");
        Object passwordObj = req.get("password");
        
        if (!(usernameObj instanceof String) || !(passwordObj instanceof String)) {
            sendJson(exchange, 401, mapOf("error", "Invalid credentials"));
            return;
        }
        
        String username = (String) usernameObj;
        String password = (String) passwordObj;
        
        User user = usersByUsername.get(username);
        if (user == null || !user.password.equals(password)) {
            sendJson(exchange, 401, mapOf("error", "Invalid credentials"));
            return;
        }
        
        String token = generateToken();
        sessions.put(token, user.id);
        
        exchange.getResponseHeaders().add("Set-Cookie", "session_id=" + token + "; Path=/; HttpOnly");
        sendJson(exchange, 200, mapOf("id", user.id, "username", user.username));
    }

    static void handleLogout(HttpExchange exchange) throws IOException {
        if (!exchange.getRequestMethod().equals("POST")) {
            sendJson(exchange, 405, mapOf("error", "Method not allowed"));
            return;
        }
        Integer userId = getUserId(exchange);
        if (userId == null) {
            sendJson(exchange, 401, mapOf("error", "Authentication required"));
            return;
        }
        
        String cookieHeader = exchange.getRequestHeaders().getFirst("Cookie");
        if (cookieHeader != null) {
            String[] cookies = cookieHeader.split(";");
            for (String cookie : cookies) {
                cookie = cookie.trim();
                if (cookie.startsWith("session_id=")) {
                    String token = cookie.substring("session_id=".length()).trim();
                    sessions.remove(token);
                }
            }
        }
        
        sendJson(exchange, 200, mapOf());
    }

    static void handleMe(HttpExchange exchange) throws IOException {
        if (!exchange.getRequestMethod().equals("GET")) {
            sendJson(exchange, 405, mapOf("error", "Method not allowed"));
            return;
        }
        Integer userId = getUserId(exchange);
        if (userId == null) {
            sendJson(exchange, 401, mapOf("error", "Authentication required"));
            return;
        }
        User user = usersById.get(userId);
        if (user == null) {
            sendJson(exchange, 401, mapOf("error", "Authentication required"));
            return;
        }
        sendJson(exchange, 200, mapOf("id", user.id, "username", user.username));
    }

    static void handlePassword(HttpExchange exchange) throws IOException {
        if (!exchange.getRequestMethod().equals("PUT")) {
            sendJson(exchange, 405, mapOf("error", "Method not allowed"));
            return;
        }
        Integer userId = getUserId(exchange);
        if (userId == null) {
            sendJson(exchange, 401, mapOf("error", "Authentication required"));
            return;
        }
        User user = usersById.get(userId);
        if (user == null) {
            sendJson(exchange, 401, mapOf("error", "Authentication required"));
            return;
        }
        
        String body = readBody(exchange);
        Map<String, Object> req;
        try {
            req = JsonParser.parseObject(body);
        } catch (Exception e) {
            sendJson(exchange, 400, mapOf("error", "Invalid request"));
            return;
        }
        
        Object oldPassObj = req.get("old_password");
        Object newPassObj = req.get("new_password");
        
        if (!(oldPassObj instanceof String) || !(newPassObj instanceof String)) {
            sendJson(exchange, 400, mapOf("error", "Invalid request"));
            return;
        }
        
        String oldPassword = (String) oldPassObj;
        String newPassword = (String) newPassObj;
        
        if (!user.password.equals(oldPassword)) {
            sendJson(exchange, 401, mapOf("error", "Invalid credentials"));
            return;
        }
        
        if (newPassword.length() < 8) {
            sendJson(exchange, 400, mapOf("error", "Password too short"));
            return;
        }
        
        user.password = newPassword;
        sendJson(exchange, 200, mapOf());
    }

    static void handleTodos(HttpExchange exchange) throws IOException {
        String method = exchange.getRequestMethod();
        if (method.equals("GET")) {
            Integer userId = getUserId(exchange);
            if (userId == null) {
                sendJson(exchange, 401, mapOf("error", "Authentication required"));
                return;
            }
            List<Map<String, Object>> result = new ArrayList<>();
            List<Todo> userTodos = new ArrayList<>();
            for (Todo todo : todos.values()) {
                if (todo.userId == userId) {
                    userTodos.add(todo);
                }
            }
            userTodos.sort(Comparator.comparingInt(t -> t.id));
            for (Todo todo : userTodos) {
                result.add(todo.toJson());
            }
            sendJson(exchange, 200, result);
        } else if (method.equals("POST")) {
            Integer userId = getUserId(exchange);
            if (userId == null) {
                sendJson(exchange, 401, mapOf("error", "Authentication required"));
                return;
            }
            
            String body = readBody(exchange);
            Map<String, Object> req;
            try {
                req = JsonParser.parseObject(body);
            } catch (Exception e) {
                sendJson(exchange, 400, mapOf("error", "Invalid request"));
                return;
            }
            
            Object titleObj = req.get("title");
            if (!(titleObj instanceof String) || ((String) titleObj).isEmpty()) {
                sendJson(exchange, 400, mapOf("error", "Title is required"));
                return;
            }
            
            String description = req.containsKey("description") && req.get("description") instanceof String 
                ? (String) req.get("description") 
                : "";
                
            int id = nextTodoId.getAndIncrement();
            String nowStr = now();
            Todo todo = new Todo();
            todo.id = id;
            todo.userId = userId;
            todo.title = (String) titleObj;
            todo.description = description;
            todo.completed = false;
            todo.createdAt = nowStr;
            todo.updatedAt = nowStr;
            
            todos.put(id, todo);
            sendJson(exchange, 201, todo.toJson());
        } else {
            sendJson(exchange, 405, mapOf("error", "Method not allowed"));
        }
    }

    static void handleTodoById(HttpExchange exchange, String idStr) throws IOException {
        int id;
        try {
            id = Integer.parseInt(idStr);
        } catch (NumberFormatException e) {
            sendJson(exchange, 404, mapOf("error", "Todo not found"));
            return;
        }
        
        Integer userId = getUserId(exchange);
        if (userId == null) {
            sendJson(exchange, 401, mapOf("error", "Authentication required"));
            return;
        }
        
        Todo todo = todos.get(id);
        if (todo == null || todo.userId != userId) {
            sendJson(exchange, 404, mapOf("error", "Todo not found"));
            return;
        }
        
        String method = exchange.getRequestMethod();
        if (method.equals("GET")) {
            sendJson(exchange, 200, todo.toJson());
        } else if (method.equals("PUT")) {
            String body = readBody(exchange);
            Map<String, Object> req;
            try {
                req = JsonParser.parseObject(body);
            } catch (Exception e) {
                sendJson(exchange, 400, mapOf("error", "Invalid request"));
                return;
            }
            
            if (req.containsKey("title")) {
                Object titleObj = req.get("title");
                if (!(titleObj instanceof String) || ((String) titleObj).isEmpty()) {
                    sendJson(exchange, 400, mapOf("error", "Title is required"));
                    return;
                }
                todo.title = (String) titleObj;
            }
            
            if (req.containsKey("description")) {
                todo.description = req.get("description") instanceof String ? (String) req.get("description") : "";
            }
            
            if (req.containsKey("completed")) {
                todo.completed = req.get("completed") instanceof Boolean ? (Boolean) req.get("completed") : false;
            }
            
            todo.updatedAt = now();
            sendJson(exchange, 200, todo.toJson());
        } else if (method.equals("DELETE")) {
            todos.remove(id);
            sendNoContent(exchange, 204);
        } else {
            sendJson(exchange, 405, mapOf("error", "Method not allowed"));
        }
    }

    static class JsonParser {
        static Map<String, Object> parseObject(String s) {
            Map<String, Object> res = new LinkedHashMap<>();
            s = s.trim();
            if (s.isEmpty() || s.equals("{}")) return res;
            if (!s.startsWith("{") || !s.endsWith("}")) throw new RuntimeException("Invalid JSON");
            s = s.substring(1, s.length() - 1).trim();
            
            int i = 0;
            while (i < s.length()) {
                while (i < s.length() && Character.isWhitespace(s.charAt(i))) i++;
                if (i >= s.length()) break;
                
                if (s.charAt(i) != '"') throw new RuntimeException("Expected '\"' at " + i);
                i++;
                int keyStart = i;
                while (i < s.length() && s.charAt(i) != '"') {
                    if (s.charAt(i) == '\\') i++;
                    i++;
                }
                String key = unescape(s.substring(keyStart, i));
                i++; 
                
                while (i < s.length() && Character.isWhitespace(s.charAt(i))) i++;
                if (i >= s.length() || s.charAt(i) != ':') throw new RuntimeException("Expected ':'");
                i++;
                
                while (i < s.length() && Character.isWhitespace(s.charAt(i))) i++;
                
                Object value;
                if (i < s.length() && s.charAt(i) == '"') {
                    i++;
                    int valStart = i;
                    while (i < s.length() && s.charAt(i) != '"') {
                        if (s.charAt(i) == '\\') i++;
                        i++;
                    }
                    value = unescape(s.substring(valStart, i));
                    i++;
                } else if (s.startsWith("true", i)) {
                    value = true;
                    i += 4;
                } else if (s.startsWith("false", i)) {
                    value = false;
                    i += 5;
                } else if (s.startsWith("null", i)) {
                    value = null;
                    i += 4;
                } else {
                    int valStart = i;
                    while (i < s.length() && s.charAt(i) != ',' && s.charAt(i) != '}') i++;
                    String numStr = s.substring(valStart, i).trim();
                    if (numStr.isEmpty()) {
                        value = null;
                    } else if (numStr.contains(".") || numStr.toLowerCase().contains("e")) {
                        value = Double.parseDouble(numStr);
                    } else {
                        value = Long.parseLong(numStr);
                    }
                }
                
                res.put(key, value);
                
                while (i < s.length() && Character.isWhitespace(s.charAt(i))) i++;
                if (i < s.length() && s.charAt(i) == ',') {
                    i++;
                }
            }
            return res;
        }
        
        static String unescape(String s) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < s.length(); i++) {
                char c = s.charAt(i);
                if (c == '\\' && i + 1 < s.length()) {
                    char next = s.charAt(i + 1);
                    if (next == '"') { sb.append('"'); i++; }
                    else if (next == '\\') { sb.append('\\'); i++; }
                    else if (next == 'n') { sb.append('\n'); i++; }
                    else if (next == 'r') { sb.append('\r'); i++; }
                    else if (next == 't') { sb.append('\t'); i++; }
                    else { sb.append(c); }
                } else {
                    sb.append(c);
                }
            }
            return sb.toString();
        }
        
        static String stringify(Object obj) {
            if (obj instanceof Map) {
                StringBuilder sb = new StringBuilder("{");
                boolean first = true;
                for (Map.Entry<?, ?> entry : ((Map<?, ?>) obj).entrySet()) {
                    if (!first) sb.append(",");
                    first = false;
                    sb.append("\"").append(escape(entry.getKey().toString())).append("\":");
                    sb.append(stringify(entry.getValue()));
                }
                sb.append("}");
                return sb.toString();
            } else if (obj instanceof List) {
                StringBuilder sb = new StringBuilder("[");
                boolean first = true;
                for (Object item : (List<?>) obj) {
                    if (!first) sb.append(",");
                    first = false;
                    sb.append(stringify(item));
                }
                sb.append("]");
                return sb.toString();
            } else if (obj instanceof String) {
                return "\"" + escape((String) obj) + "\"";
            } else if (obj instanceof Boolean) {
                return obj.toString();
            } else if (obj instanceof Number) {
                return obj.toString();
            } else if (obj == null) {
                return "null";
            } else {
                return "\"" + escape(obj.toString()) + "\"";
            }
        }
        
        static String escape(String s) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < s.length(); i++) {
                char c = s.charAt(i);
                if (c == '"') sb.append("\\\"");
                else if (c == '\\') sb.append("\\\\");
                else if (c == '\n') sb.append("\\n");
                else if (c == '\r') sb.append("\\r");
                else if (c == '\t') sb.append("\\t");
                else sb.append(c);
            }
            return sb.toString();
        }
    }

    public static void main(String[] args) throws Exception {
        int port = 8080;
        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--port") && i + 1 < args.length) {
                port = Integer.parseInt(args[i + 1]);
            }
        }
        
        HttpServer server = HttpServer.create(new InetSocketAddress("0.0.0.0", port), 0);
        
        server.createContext("/register", Main::handleRegister);
        server.createContext("/login", Main::handleLogin);
        server.createContext("/logout", Main::handleLogout);
        server.createContext("/me", Main::handleMe);
        server.createContext("/password", Main::handlePassword);
        server.createContext("/todos", exchange -> {
            String path = exchange.getRequestURI().getPath();
            if (path.equals("/todos")) {
                if (exchange.getRequestMethod().equals("GET")) handleTodos(exchange);
                else if (exchange.getRequestMethod().equals("POST")) handleTodos(exchange);
                else sendJson(exchange, 405, mapOf("error", "Method not allowed"));
            } else if (path.startsWith("/todos/")) {
                String idStr = path.substring(7);
                handleTodoById(exchange, idStr);
            } else {
                sendJson(exchange, 404, mapOf("error", "Not found"));
            }
        });
        
        server.setExecutor(Executors.newFixedThreadPool(10));
        server.start();
        System.out.println("Server started on port " + port);
    }
}

File: data_baseline/openrouter__qwen__qwen3.7-plus/rust/rep2/src/main.rs
use axum::{
    routing::{get, post, put},
    Router,
    extract::{State, Path},
    http::StatusCode,
    response::{IntoResponse, Json},
};
use axum_extra::extract::cookie::{Cookie, CookieJar};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use std::sync::atomic::{AtomicU32, Ordering};
use chrono::Utc;
use regex::Regex;
use uuid::Uuid;

#[derive(Clone, Serialize, Deserialize)]
struct User {
    id: u32,
    username: String,
    password: String,
}

#[derive(Clone, Serialize)]
struct Todo {
    id: u32,
    #[serde(skip)]
    user_id: u32,
    title: String,
    description: String,
    completed: bool,
    created_at: String,
    updated_at: String,
}

struct AppState {
    users: RwLock<HashMap<u32, User>>,
    username_to_id: RwLock<HashMap<String, u32>>,
    sessions: RwLock<HashMap<String, u32>>,
    todos: RwLock<HashMap<u32, Todo>>,
    user_todos: RwLock<HashMap<u32, Vec<u32>>>,
    next_user_id: AtomicU32,
    next_todo_id: AtomicU32,
}

impl AppState {
    fn new() -> Self {
        Self {
            users: RwLock::new(HashMap::new()),
            username_to_id: RwLock::new(HashMap::new()),
            sessions: RwLock::new(HashMap::new()),
            todos: RwLock::new(HashMap::new()),
            user_todos: RwLock::new(HashMap::new()),
            next_user_id: AtomicU32::new(1),
            next_todo_id: AtomicU32::new(1),
        }
    }
}

#[derive(Deserialize)]
struct RegisterReq {
    username: String,
    password: String,
}

#[derive(Deserialize)]
struct LoginReq {
    username: String,
    password: String,
}

#[derive(Deserialize)]
struct ChangePasswordReq {
    old_password: String,
    new_password: String,
}

#[derive(Deserialize)]
struct CreateTodoReq {
    title: Option<String>,
    description: Option<String>,
}

#[derive(Deserialize)]
struct UpdateTodoReq {
    title: Option<String>,
    description: Option<String>,
    completed: Option<bool>,
}

fn get_user_id(jar: &CookieJar, state: &Arc<AppState>) -> Result<u32, (StatusCode, Json<serde_json::Value>)> {
    if let Some(cookie) = jar.get("session_id") {
        let sessions = state.sessions.read().unwrap();
        if let Some(&user_id) = sessions.get(cookie.value()) {
            return Ok(user_id);
        }
    }
    Err((StatusCode::UNAUTHORIZED, Json(serde_json::json!({"error": "Authentication required"}))))
}

async fn register(
    State(state): State<Arc<AppState>>,
    Json(req): Json<RegisterReq>,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let username_regex = Regex::new(r"^[a-zA-Z0-9_]+$").unwrap();
    if req.username.len() < 3 || req.username.len() > 50 || !username_regex.is_match(&req.username) {
        return Err((StatusCode::BAD_REQUEST, Json(serde_json::json!({"error": "Invalid username"}))));
    }
    if req.password.len() < 8 {
        return Err((StatusCode::BAD_REQUEST, Json(serde_json::json!({"error": "Password too short"}))));
    }

    let mut username_to_id = state.username_to_id.write().unwrap();
    if username_to_id.contains_key(&req.username) {
        return Err((StatusCode::CONFLICT, Json(serde_json::json!({"error": "Username already exists"}))));
    }

    let user_id = state.next_user_id.fetch_add(1, Ordering::SeqCst);
    let user = User {
        id: user_id,
        username: req.username.clone(),
        password: req.password,
    };

    state.users.write().unwrap().insert(user_id, user);
    let username_for_map = req.username.clone();
    username_to_id.insert(username_for_map, user_id);

    Ok((StatusCode::CREATED, Json(serde_json::json!({
        "id": user_id,
        "username": req.username
    }))))
}

async fn login(
    State(state): State<Arc<AppState>>,
    jar: CookieJar,
    Json(req): Json<LoginReq>,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let user_info = {
        let username_to_id = state.username_to_id.read().unwrap();
        if let Some(&user_id) = username_to_id.get(&req.username) {
            let users = state.users.read().unwrap();
            if let Some(user) = users.get(&user_id) {
                if user.password == req.password {
                    Some((user_id, user.username.clone()))
                } else {
                    None
                }
            } else {
                None
            }
        } else {
            None
        }
    };

    if let Some((user_id, username)) = user_info {
        let token = Uuid::new_v4().to_string();
        state.sessions.write().unwrap().insert(token.clone(), user_id);
        
        let cookie = Cookie::build(("session_id", token))
            .path("/")
            .http_only(true)
            .build();
            
        Ok((jar.add(cookie), (StatusCode::OK, Json(serde_json::json!({
            "id": user_id,
            "username": username
        })))))
    } else {
        Err((StatusCode::UNAUTHORIZED, Json(serde_json::json!({"error": "Invalid credentials"}))))
    }
}

async fn logout(
    State(state): State<Arc<AppState>>,
    jar: CookieJar,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let _user_id = get_user_id(&jar, &state)?;
    
    if let Some(cookie) = jar.get("session_id") {
        state.sessions.write().unwrap().remove(cookie.value());
    }
    
    Ok((jar.remove(Cookie::from("session_id")), (StatusCode::OK, Json(serde_json::json!({})))))
}

async fn me(
    State(state): State<Arc<AppState>>,
    jar: CookieJar,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let user_id = get_user_id(&jar, &state)?;
    let users = state.users.read().unwrap();
    if let Some(user) = users.get(&user_id) {
        Ok((StatusCode::OK, Json(serde_json::json!({
            "id": user.id,
            "username": user.username.clone()
        }))))
    } else {
        Err((StatusCode::UNAUTHORIZED, Json(serde_json::json!({"error": "Authentication required"}))))
    }
}

async fn change_password(
    State(state): State<Arc<AppState>>,
    jar: CookieJar,
    Json(req): Json<ChangePasswordReq>,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let user_id = get_user_id(&jar, &state)?;
    
    if req.new_password.len() < 8 {
        return Err((StatusCode::BAD_REQUEST, Json(serde_json::json!({"error": "Password too short"}))));
    }
    
    let mut users = state.users.write().unwrap();
    if let Some(user) = users.get_mut(&user_id) {
        if user.password != req.old_password {
            return Err((StatusCode::UNAUTHORIZED, Json(serde_json::json!({"error": "Invalid credentials"}))));
        }
        user.password = req.new_password;
        Ok((StatusCode::OK, Json(serde_json::json!({}))))
    } else {
        Err((StatusCode::UNAUTHORIZED, Json(serde_json::json!({"error": "Authentication required"}))))
    }
}

async fn get_todos(
    State(state): State<Arc<AppState>>,
    jar: CookieJar,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let user_id = get_user_id(&jar, &state)?;
    
    let todo_ids = {
        let user_todos = state.user_todos.read().unwrap();
        user_todos.get(&user_id).cloned().unwrap_or_default()
    };
    
    let todos = state.todos.read().unwrap();
    let mut result: Vec<Todo> = todo_ids
        .into_iter()
        .filter_map(|id| todos.get(&id).cloned())
        .collect();
        
    result.sort_by_key(|t| t.id);
    
    Ok((StatusCode::OK, Json(serde_json::to_value(result).unwrap())))
}

async fn create_todo(
    State(state): State<Arc<AppState>>,
    jar: CookieJar,
    Json(req): Json<CreateTodoReq>,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let user_id = get_user_id(&jar, &state)?;
    
    let title = req.title.unwrap_or_default();
    if title.trim().is_empty() {
        return Err((StatusCode::BAD_REQUEST, Json(serde_json::json!({"error": "Title is required"}))));
    }
    
    let description = req.description.unwrap_or_default();
    let now = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();
    
    let todo_id = state.next_todo_id.fetch_add(1, Ordering::SeqCst);
    let todo = Todo {
        id: todo_id,
        user_id,
        title,
        description,
        completed: false,
        created_at: now.clone(),
        updated_at: now,
    };
    
    state.todos.write().unwrap().insert(todo_id, todo.clone());
    
    let mut user_todos = state.user_todos.write().unwrap();
    user_todos.entry(user_id).or_default().push(todo_id);
    
    Ok((StatusCode::CREATED, Json(serde_json::to_value(todo).unwrap())))
}

async fn get_todo(
    State(state): State<Arc<AppState>>,
    jar: CookieJar,
    Path(todo_id): Path<u32>,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let user_id = get_user_id(&jar, &state)?;
    
    let todos = state.todos.read().unwrap();
    if let Some(todo) = todos.get(&todo_id) {
        if todo.user_id == user_id {
            return Ok((StatusCode::OK, Json(serde_json::to_value(todo).unwrap())));
        }
    }
    Err((StatusCode::NOT_FOUND, Json(serde_json::json!({"error": "Todo not found"}))))
}

async fn update_todo(
    State(state): State<Arc<AppState>>,
    jar: CookieJar,
    Path(todo_id): Path<u32>,
    Json(req): Json<UpdateTodoReq>,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let user_id = get_user_id(&jar, &state)?;
    
    let mut todos = state.todos.write().unwrap();
    if let Some(todo) = todos.get_mut(&todo_id) {
        if todo.user_id != user_id {
            return Err((StatusCode::NOT_FOUND, Json(serde_json::json!({"error": "Todo not found"}))));
        }
        
        if let Some(title) = req.title {
            if title.trim().is_empty() {
                return Err((StatusCode::BAD_REQUEST, Json(serde_json::json!({"error": "Title is required"}))));
            }
            todo.title = title;
        }
        
        if let Some(desc) = req.description {
            todo.description = desc;
        }
        
        if let Some(completed) = req.completed {
            todo.completed = completed;
        }
        
        todo.updated_at = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();
        
        return Ok((StatusCode::OK, Json(serde_json::to_value(todo.clone()).unwrap())));
    }
    
    Err((StatusCode::NOT_FOUND, Json(serde_json::json!({"error": "Todo not found"}))))
}

async fn delete_todo(
    State(state): State<Arc<AppState>>,
    jar: CookieJar,
    Path(todo_id): Path<u32>,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let user_id = get_user_id(&jar, &state)?;
    
    let mut todos = state.todos.write().unwrap();
    if let Some(todo) = todos.get(&todo_id) {
        if todo.user_id != user_id {
            return Err((StatusCode::NOT_FOUND, Json(serde_json::json!({"error": "Todo not found"}))));
        }
        
        todos.remove(&todo_id);
        
        let mut user_todos = state.user_todos.write().unwrap();
        if let Some(list) = user_todos.get_mut(&user_id) {
            list.retain(|&id| id != todo_id);
        }
        
        return Ok(StatusCode::NO_CONTENT);
    }
    
    Err((StatusCode::NOT_FOUND, Json(serde_json::json!({"error": "Todo not found"}))))
}

#[tokio::main]
async fn main() {
    let args: Vec<String> = std::env::args().collect();
    let mut port = 8080;
    let mut i = 1;
    while i < args.len() {
        if args[i] == "--port" && i + 1 < args.len() {
            port = args[i + 1].parse().unwrap_or(8080);
            i += 2;
        } else {
            i += 1;
        }
    }
    
    let app = Router::new()
        .route("/register", post(register))
        .route("/login", post(login))
        .route("/logout", post(logout))
        .route("/me", get(me))
        .route("/password", put(change_password))
        .route("/todos", get(get_todos).post(create_todo))
        .route("/todos/:id", get(get_todo).put(update_todo).delete(delete_todo))
        .with_state(Arc::new(AppState::new()));

    let listener = tokio::net::TcpListener::bind(format!("0.0.0.0:{}", port)).await.unwrap();
    println!("Listening on 0.0.0.0:{}", port);
    axum::serve(listener, app).await.unwrap();
}
